Collocated processes instructions (#13224)
apache/druid: docs/development/extensions-contrib/prometheus.md

This extension exposes [Druid metrics](https://druid.apache.org/docs/latest/operations/metrics.html) for collection by a Prometheus server (https://prometheus.io/).

Emitter is enabled by setting `druid.emitter=prometheus` [configs](https://druid.apache.org/docs/latest/configuration/index.html#enabling-metrics) or include `prometheus` in the composing emitter list. 

> In certain instances, Druid processes may be colocated on the same host. For example, the Broker and Router may share the same server. Other colocated processes include the Historical and MiddleManager or the Coordinator and Overlord. When you have colocated processes, specify `druid.emitter.prometheus.port` separately for each process on each host. For example, even if the Broker and Router share the same host, the Broker runtime properties and the Router runtime properties each need to list `druid.emitter.prometheus.port`, and the port value for both must be different. While peon tasks are typically colocated with MiddleManagers and often Historicals, peons should use `pushgateway` for `druid.emitter.prometheus.strategy` and should not start an HTTP server to push metrics (they are short lived batch processes and multiple of these processes can be running at once) . 

Reviewer: ```suggestion
In certain instances, Druid processes may be colocated on the same host. For example, the Broker and Router may share the same server. Other colocated processes include the Historical and MiddleManager or the Coordinator and Overlord. When you have colocated processes, specify `druid.emitter.prometheus.port` separately for each process on each host. For example, even if the Broker and Router share the same host, the Broker runtime properties and the Router runtime properties each need to list `druid.emitter.prometheus.port`, and the port value for both must be different.

Peon tasks are short-lived batch processes typically colocated with MiddleManagers and often Historicals. Multiple peons may run at once. These tasks should use `pushgateway` for `druid.emitter.prometheus.strategy` and should not start an HTTP server to push metrics. 
```
There's enough material for peon tasks to make this subject its own paragraph (rearranged order also)

> To configure push gateway for peons you can add somehting like this in you middlemanager runtime properties file:

Reviewer: ```suggestion
To configure Pushgateway for peons, pass the Prometheus options to the peon's JVM by setting the [`javaOptsArray` property](../../configuration/index.md#middlemanager-configuration) in your MiddleManager runtime properties file:
```
Using same spelling/style as [Prometheus docs](https://prometheus.io/docs/practices/pushing/)

> ```
> druid.indexer.runner.javaOptsArray=["druid.emitter.prometheus.strategy=pushgateway","druid.emitter.prometheus.pushGatewayAddress=http://pushgatewayhost:pushgatewayport"]
> ```

Reviewer: `````suggestion
```
druid.indexer.runner.javaOptsArray=["druid.emitter.prometheus.strategy=pushgateway","druid.emitter.prometheus.pushGatewayAddress=http://PUSHGATEWAYHOST:PUSHGATEWAYPORT"]
```
Replace `PUSHGATEWAYHOST:PUSHGATEWAYPORT` with your Pushgateway address.
`````

Reviewer: Adds the additional "replace" instruction so someone who's quickly copying and pasting knows what they need to change.

This extension exposes [Druid metrics](https://druid.apache.org/docs/latest/operations/metrics.html) for collection by a Prometheus server (https://prometheus.io/).

Emitter is enabled by setting `druid.emitter=prometheus` [configs](https://druid.apache.org/docs/latest/configuration/index.html#enabling-metrics) or include `prometheus` in the composing emitter list. 

> In certain instances, Druid processes may be colocated on the same host. For example, the Broker and Router may share the same server. Other colocated processes include the Historical and MiddleManager or the Coordinator and Overlord. When you have colocated processes, specify `druid.emitter.prometheus.port` separately for each process on each host. For example, even if the Broker and Router share the same host, the Broker runtime properties and the Router runtime properties each need to list `druid.emitter.prometheus.port`, and the port value for both must be different.

Reviewer 2: I'm suggesting to move this paragraph after the configuration section. Because I think it's a little strange for people to see the configuration for port here before they get a full picture of the whole configuration of this extension.

Author: Thanks for the feedback,  I've made the changes requested.

> Peon tasks are short-lived batch processes typically colocated with MiddleManagers and often Historicals. Multiple peons may run at once. These tasks should use `pushgateway` for `druid.emitter.prometheus.strategy` and should not start an HTTP server to push metrics. 

Reviewer 2: There's a section below the configuration to describe the configuration for peon tasks. 
I think this paragraph is not necessary.

Emitter is enabled by setting `druid.emitter=prometheus` [configs](https://druid.apache.org/docs/latest/configuration/index.html#emitting-metrics) or include `prometheus` in the composing emitter list. 

> Note that if you're colocating druid processes on the same host(such as broker/router, historical/middlemanager, coordinator/overlord), then the druid.emitter.prometheus.port would need to be separately specified in each of the runtime properties configuration for each host and be different between collocated processes. Additionally peons should emit metrics via push gateway. 

Reviewer: A few comments:
* **Can we break up the first sentence for better readability. Something like:** In certain instances, Druid processes may be colocated on the same host. For example, the Broker and Router may share the same server. Other colocated processes include the Historical and MiddleManager or the Coordinator and Overlord.
* **Code format for properties:** `druid.emitter.prometheus.port`
* **Prefer active voice:** Specify `druid.emitter.prometheus.port` separately ...
* **Is there a better transition between the colocation and peon tasks? Should these be in separate paragraph, or do they cover the same concept related to colocated processes?**

Most of the above are in the following suggestion. 
```suggestion
In certain instances, Druid processes may be colocated on the same host. For example, the Broker and Router may share the same server. Other colocated processes include the Historical and MiddleManager or the Coordinator and Overlord. When you have colocated processes, specify `druid.emitter.prometheus.port` separately for each process on each host. For example, even if the Broker and Router share the same host, the Broker runtime properties and the Router runtime properties each need to list `druid.emitter.prometheus.port`, and the port value for both must be different. Peon tasks should use `pushgateway` for `druid.emitter.prometheus.strategy`.
```

| `druid.emitter.prometheus.pushGatewayAddress` | Pushgateway address. Required if using `pushgateway` strategy.                                                                                                                                                                         | no        | none                                 |
|`druid.emitter.prometheus.flushPeriod`|Emit metrics to Pushgateway every `flushPeriod` seconds. Required if `pushgateway` strategy is used.|no|15|

> ### Different ports for colocated druid processes

Reviewer: ```suggestion
### Ports for colocated Druid processes
```

In certain instances, Druid processes may be colocated on the same host. For example, the Broker and Router may share the same server. Other colocated processes include the Historical and MiddleManager or the Coordinator and Overlord. When you have colocated processes, specify `druid.emitter.prometheus.port` separately for each process on each host. For example, even if the Broker and Router share the same host, the Broker runtime properties and the Router runtime properties each need to list `druid.emitter.prometheus.port`, and the port value for both must be different.

### Override properties for Peon Tasks

Peon tasks are created dynamically by middle managers and have dynamic host and port addresses. Since the `exporter` strategy allows Prometheus to read only from a fixed address, it cannot be used for peon tasks.
So, these tasks need to be configured to use `pushgateway` strategy to push metrics from Druid to prometheus gateway.

If this emitter is configured to use `exporter` strategy globally, some of the above configurations need to be overridden in the middle manager so that spawned peon tasks can still use the `pushgateway` strategy.

> 

Reviewer 2: ```suggestion

```

Extra change
